### DISCUSSION TITLE
Aiming zero allocation with CancellationTokenSource

### DISCUSSION THREAD

**Discussion by Begounet** [Cysharp/UniTask#601] [Q&A]
Hi!

UniTask is really awesome to have reached successfully zero alloc, and I would like to stick on this way by having zero allocation with the use of CancellationTokenSource. It is kind of really important since it has its place in most of methods using UniTask.

I think using a pool could be great. However, since there is no way to reset the CancellationTokenSource ([TryReset](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.tryreset?view=net-8.0) not being accessible in Unity), I would like to know if someone has found a way to get around this.

Thank you!

**Comment by neuecc** [comment]
TryReset resets the internal Timer, and if Canceled, TryReset will not succeed and cannot be reused.
If you really want to reset the timer, you can replace TryReset by calling CancelAfter(Timeout.InfiniteTimeSpan).
This is a hack that was used before .NET 6.

**Reply by Begounet** [reply]
Oh okay! I have seen this hack somewhere but it did not seem to work because I was thinking that it was resetting the "IsCancellationRequested", not only the timer! Thank you for this enlightenment!

However, it would not be enough to be reused so I could not use it a standard CancellationTokenSource pool...

Just looking at the code source with ILSpy (in case it would be possible to reset at least by reflection -- not even sure if it is possible to do it without allocation...) but the class seems pretty complex and I am not sure I could just reset the field like this.

Wondering... Should I recreate my own (way more simple) CancellationTokenSource system seems doable? Maybe by doing something really similar to "AttachExternalCancellationSource"?

**Comment by timcassell** [comment]
Even in .Net 6, TryReset fails to reset the cancelation if the source was already canceled.

I wrote my own CancelationToken that uses object pooling to achieve zero alloc. Check it out in my [ProtoPromise library](https://github.com/timcassell/ProtoPromise).

**Reply by Begounet** [reply]
Thank you for your recommendation! I didn't know such library exists!
I am used to UniTask and really like it but if I cannot find a way around, I will definitively checkout your library!

**Comment by BowserKingKoopa** [comment]
I'm kind of stuck here as well.  I'm using UniTask for my first big Unity project and everything is going swimmingly until you need to cancel/reset something.  Is there no way to do cancellation allocation free?
